### PR TITLE
VS-286 Support mixed-mode local bindings for Vectorize in Miniflare

### PR DESCRIPTION
**Pull request by garrettgu10**
## What this PR solves / how to test

Fixes VS-286.

## Author has addressed the following

- Tests
  - [ ] TODO (before merge)
  - [x] Tests included
  - [ ] Tests not necessary because:
- E2E Tests CI Job required? (Use "e2e" label or ask maintainer to run separately)
  - [ ] I don't know
  - [x] Required
  - [ ] Not required because:
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [ ] TODO (before merge)
  - [x] Changeset included
  - [ ] Changeset not necessary because:
- Public documentation
  - [ ] TODO (before merge)
  - [ ] Cloudflare docs PR(s): 
  - [x] Documentation not necessary because: I don't think this limitation is mentioned anywhere in docs and it's generally expected that bindings work in local dev mode.

### CODE REVIEW

**Comment by cloudkite** [comment]
What’s the reasoning behind disabling write in local mode? Doesn’t seem very useful without ability to write

**Comment by garrettgu10** [comment]
@cloudkite Both D1 and R2 operate on local instances that are separate from prod. Since we're directly connecting to the prod Vectorize instance instead, we definitely don't want to catch people off guard and make them cause changes in prod while testing "locally".

We're considering re-enabling write operations behind an optional flag, and I think that's probably the best option.

**Comment by cloudkite** [comment]
@garrettgu10 makes sense in terms of accidentally writing to a live db instance. However having trouble imaging how a readonly "local" db will be useful for development/testing purposes?

**Comment by garrettgu10** [comment]
@cloudkite we are not (yet) launching a local simulation database for vectorize. All requests are sent to the prod database in local dev mode.

**Comment by cloudkite** [comment]
@garrettgu10 yup I understood that part. I just meant if I can’t write I’ll still need to use —remote

How about when wrangler is in local mode it refuses to bind unless the index name has suffix of “-local-dev” or similar, that would make it very unlikely for someone to accidentally bind to their live/prod vectorize index

File: packages/wrangler/src/vectorize/fetcher.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+				{ status: 403 }
+			);
+		}
+
+		const url = request.url.replace(
+			"http://vector-search/",
+			`/accounts/${accountId}/vectorize/${indexVersion}/indexes/${indexId}/`
+		);
+
+		// TODO: v1 endpoints have a different format
```
Comment (sejoker) [inline packages/wrangler/src/vectorize/fetcher.ts:57]: we won't support v1 endpoints as v1 is EOL.

**Review by sejoker — APPROVED** [review]

**Review by irvinebroque — COMMENTED** [review]
Let's make sure to update these docs when this lands:

https://developers.cloudflare.com/workers/testing/local-development/#supported-resource-bindings-in-different-environments

**Comment by garrettgu10** [comment]
@mrbbot Thanks for your comments, they've been super helpful internally and we've discussed them at length. I agree that a local simulator is the superior solution re. DX and seems fairly simple to implement as a first-pass, but the team doesn't have the bandwidth for maintaining such a solution at the moment, esp. wrt. feature-parity with updates to the service. 

With that in mind, I think the best compromise is to allow users to bind to production indexes through an optional flag such as `--experimental-vectorize-bind-to-production`. That way, users must explicitly opt-in, minimizing possibility for confusion. Then, sessions launched without the flag will have no Vectorize bindings until we're ready to launch a local simulator. That's the approach we're going with for now.

File: packages/wrangler/src/dev/miniflare.ts [hunk 2]
```diff
@@ -598,6 +603,36 @@ export function buildMiniflareBindingOptions(config: MiniflareBindingsConfig): {
 		};
 	}
 
+	if (bindings.vectorize) {
+		for (const vectorizeBinding of bindings.vectorize) {
+			const bindingName = vectorizeBinding.binding;
+			const indexName = vectorizeBinding.index_name;
+			const indexVersion = "v2"; // TODO: support v1?
```
[the remark this answers is not in the document]
Comment (penalosa) [reply packages/wrangler/src/dev/miniflare.ts:610]: Is there any way to warn/error if a user tries to use a v1 binding?

File: packages/wrangler/src/dev/miniflare.ts [hunk 3]
```diff
@@ -889,9 +924,8 @@ export async function buildMiniflareOptions(
 	if (config.bindings.vectorize?.length) {
 		if (!didWarnMiniflareVectorizeSupport) {
 			didWarnMiniflareVectorizeSupport = true;
-			// TODO: add local support for Vectorize bindings (https://github.com/cloudflare/workers-sdk/issues/4360)
 			logger.warn(
-				"Vectorize bindings are not currently supported in local mode. Please use --remote if you are working with them."
+				"Using Vectorize always accesses your Cloudflare account in order to run queries and may incur usage charges even in local development. Write operations are also not enabled in local development; use `--remote` if write operations are needed."
```
Comment (penalosa) [inline packages/wrangler/src/dev/miniflare.ts:928]: Why are write operations not enabled?

**Comment by penalosa** [comment]
> With that in mind, I think the best compromise is to allow users to bind to production indexes through an optional flag such as `--experimental-vectorize-bind-to-production`. That way, users must explicitly opt-in, minimizing possibility for confusion. Then, sessions launched without the flag will have no Vectorize bindings until we're ready to launch a local simulator. That's the approach we're going with for now.

@garrettgu10 This feels like a pretty big break from how other bindings are implemented, and I'd be concerned about adding an `--experimental-vectorize-bind-to-production` flag without a larger discussion of how that fits in to other bindings. This PR looks relatively reasonable as a mixed mode implementation for Vectorize (other than the write limitations, which I touched on in another comment), but I'd been assuming that there were fundamental limitations that required mixed mode for Vectorize (similar to Workers AI & Browser bindings, both of which require large downloads/specific hardware).

If there aren't any fundamental limits on getting Vectorize working locally, would it be possible to revisit that discussion? Do you have a link to a decision doc or anything around this?

File: packages/wrangler/e2e/dev-with-resources.test.ts [hunk 4]
```diff
@@ -579,7 +640,6 @@ describe.sequential.each(RUNTIMES)("Bindings: $flags", ({ runtime, flags }) => {
 	it.skipIf(isLocal).todo("exposes send email bindings");
 	it.skipIf(isLocal).todo("exposes browser bindings");
 	it.skipIf(isLocal).todo("exposes Workers AI bindings");
-	it.skipIf(isLocal).todo("exposes Vectorize bindings");
```
Comment (penalosa) [inline packages/wrangler/e2e/dev-with-resources.test.ts:582]: 🎉

**Pull request merged**